File: 류호석배 알고리즘 코딩 테스트/제2회/3번-짠돌이 호석/solution.cpp
#include <iostream>
#include <algorithm>
#include <string>
#define NM 105
#define INF 0x7fffffff
#define vi vector<int>
FILE* in = stdin, * out = stdout;
using namespace std;

int n1, m1, n2, m2;
char a[NM][NM], b[NM][NM], temp[NM][NM];
int ans = INF;

void input() {
    cin >> n1 >> m1;
    for (int i = 1; i <= n1; i++) for (int j = 1; j <= m1; j++) cin >> a[i][j];
    cin >> n2 >> m2;
    for (int i = 1; i <= n2; i++) for (int j = 1; j <= m2; j++) cin >> b[i][j];
}
void rotate(char a[][NM], int& n, int& m) {
    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= m; j++) {
            temp[j][n - i + 1] = a[i][j];
        }
    }
    swap(n, m);
    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= m; j++) {
            a[i][j] = temp[i][j];
        }
    }
}
bool possible(int shift_row, int shift_col) {
    for (int ai = 1; ai <= n1; ai++) {
        for (int aj = 1; aj <= m1; aj++) {
            if (a[ai][aj] == '0')
                continue;
            int bi = ai + shift_row, bj = aj + shift_col;
            if (1 <= bi && bi <= n2 && 1 <= bj && bj <= m2 && b[bi][bj] == '1')
                return false;
        }
    }
    return true;
}
void pro() {
    for (int rot = 1; rot <= 4; rot++) {
        rotate(a, n1, m1);
        for (int shift_row = -51; shift_row <= 51; shift_row++) {
            for (int shift_col = -51; shift_col <= 51; shift_col++) {
                if (possible(shift_row, shift_col)) {
                    int row = max(n2 - 1, shift_row + n1 - 1) - min(0, shift_row) + 1;
                    int col = max(m2 - 1, shift_col + m1 - 1) - min(0, shift_col) + 1;
                    ans = min(ans, row * col);
                }
            }
        }
    }
    cout << ans;
}

int main() {
    ios::sync_with_stdio(false); cin.tie(NULL);
    input();
    pro();
    return 0;
}

File: 류호석배 알고리즘 코딩 테스트/제2회/3번-짠돌이 호석/solution.py
import sys

n1, m1 = map(int, sys.stdin.readline().strip().split())
pict1 = []
for i in range(n1):
	pict1.append(sys.stdin.readline().strip())

n2, m2 = map(int, sys.stdin.readline().strip().split())
pict2 = []
for i in range(n2):
	pict2.append(sys.stdin.readline().strip())

res = 40404
for _ in range(4):
	temp = [[0 for j in range(n2)] for i in range(m2)]
	for i in range(n2):
		for j in range(m2):
			temp[j][n2 - i - 1] = pict2[i][j]
	pict2 = temp
	n2, m2 = m2, n2
	for i in range(-51, 51):
		for j in range(-51, 51):
			pos = True
			for x in range(n1):
				for y in range(m1):
					if 0 <= x + i < n2 and 0 <= y + j < m2 and pict1[x][y] == '1' and pict2[x + i][y + j] == '1':
						pos = False
						break
				if not pos:
					break
			if pos:
				res = min(res, (max(n1 + i, n2) - min(0, i)) * (max(m1 + j, m2) - min(0, j)))
print(res)